File: src/lib/desktop/menu.ts
/**
 * Native menu events, delivered to whichever component owns the action.
 *
 * Rust emits one `menu` event carrying the item id; a single listener fans it
 * out here. Components register the ids they can handle, so the same actions
 * back both the menu bar and the on-screen controls with no second code path.
 */
import { useEffect, useRef } from 'react';
import { listen } from '@tauri-apps/api/event';
import { invoke } from '@tauri-apps/api/core';

export type MenuId =
  | 'open-game'
  | 'library'
  | 'save'
  | 'restore'
  | 'restart'
  | 'achievements'
  | 'settings'
  | 'toggle-sidebar'
  | 'toggle-panel'
  | 'palette'
  | 'stats'
  | 'toggle-focus'
  | 'zoom-in'
  | 'zoom-out'
  | 'zoom-reset';

type Handler = () => void;

const handlers = new Map<MenuId, Set<Handler>>();
let listening = false;

/**
 * The same ids, from the keyboard.
 *
 * Native accelerators alone were not enough: a menu that fails to build takes
 * every one of its shortcuts with it silently, and a webview can swallow
 * combinations before the menu ever sees them. Handling the keys here as well
 * means the shortcuts work whether or not the native menu is behaving, and
 * both routes end at the same handler.
 */
const KEYS: Record<string, MenuId> = {
  'mod+o': 'open-game',
  'mod+shift+l': 'library',
  'mod+s': 'save',
  'mod+l': 'restore',
  'mod+shift+r': 'restart',
  'mod+shift+a': 'achievements',
  'mod+shift+s': 'stats',
  'mod+,': 'settings',
  'mod+\\': 'toggle-sidebar',
  'mod+i': 'toggle-panel',
  'mod+k': 'palette',
  'mod+shift+f': 'toggle-focus',
  'mod+=': 'zoom-in',
  'mod++': 'zoom-in',
  'mod+-': 'zoom-out',
  'mod+0': 'zoom-reset',
};

function fire(id: MenuId) {
  const bound = handlers.get(id);
  if (!bound?.size) return false;
  for (const fn of bound) fn();
  return true;
}

function combo(e: KeyboardEvent): string | null {
  if (!e.ctrlKey && !e.metaKey) return null;
  const key = e.key.length === 1 ? e.key.toLowerCase() : e.key.toLowerCase();
  return `mod+${e.shiftKey ? 'shift+' : ''}${key}`;
}

function start() {
  if (listening) return;
  listening = true;

  void listen<string>('menu', (event) => {
    fire(event.payload as MenuId);
  });

  document.addEventListener(
    'keydown',
    (e) => {
      const key = combo(e);
      const id = key && KEYS[key];
      if (!id) return;
      /* Only swallow the key if something was actually listening for it —
         otherwise Ctrl+S on the library page would block nothing and do
         nothing, which reads as a broken shortcut. */
      if (fire(id)) e.preventDefault();
    },
    true,
  );
}

/**
 * Grey out the game-only items when no game is open.
 *
 * Leaving Save and Restart enabled on the library page offered actions that
 * could not work; the menu is the one place a reader looks to find out what is
 * possible right now.
 */
/**
 * Hide the whole menu bar for focus mode.
 *
 * It is part of the window rather than the page, so CSS cannot reach it. On
 * macOS the system menu bar hides itself in fullscreen and this is a no-op.
 */
export async function setMenuVisible(visible: boolean): Promise<void> {
  try {
    await invoke('set_menu_visible', { visible });
  } catch {
    /* nothing to do: the menu simply stays where it is */
  }
}

export async function setGameMenuEnabled(enabled: boolean): Promise<void> {
  try {
    await invoke('set_game_menu_enabled', { enabled });
  } catch {
    /* an older build without the command simply keeps them enabled */
  }
}

/**
 * Bind menu ids to actions for as long as the component is mounted.
 *
 * The map is read through a ref, so a component can pass a fresh object every
 * render — which it will, since these close over current state — without
 * resubscribing.
 */
export function useMenu(map: Partial<Record<MenuId, Handler>>) {
  const latest = useRef(map);
  latest.current = map;

  useEffect(() => {
    start();

    const ids = Object.keys(latest.current) as MenuId[];
    const bound = ids.map((id) => {
      const fn: Handler = () => latest.current[id]?.();
      if (!handlers.has(id)) handlers.set(id, new Set());
      handlers.get(id)!.add(fn);
      return [id, fn] as const;
    });

    return () => {
      for (const [id, fn] of bound) handlers.get(id)?.delete(fn);
    };
    // Ids are stable for a given component; the actions are read through the ref.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [Object.keys(map).join(',')]);
}

File: src-tauri/src/menu.rs
//! The native menu bar.
//!
//! Every item forwards its id to the front end as a `menu` event rather than
//! acting in Rust: the actions belong to the engine, which lives in the
//! webview. Rust owns only the shape of the menu and the accelerators.
//!
//! One accelerator is deliberately absent. `CmdOrCtrl+R` is reload in every
//! webview, and a player who meant "restart the chapter" would instead drop
//! the whole session, so restart is bound to `CmdOrCtrl+Shift+R`.
//!
//! Symbol keys are written as key codes — `Equal`, `Minus`, `Digit0`,
//! `Backslash`, `Comma` — not as the characters they produce. `Plus` and `,`
//! are not parseable accelerators, and one unparseable string fails the whole
//! menu, which takes every other shortcut down with it silently. The front end
//! also handles these combinations itself (lib/desktop/menu.ts), so a menu that
//! fails to build no longer means a keyboard that does nothing.

use tauri::menu::{Menu, MenuItemBuilder, PredefinedMenuItem, SubmenuBuilder};
/* Only the application menu uses this, and that menu is macOS-only. Importing
   it unconditionally warns on every Windows and Linux build. */
#[cfg(target_os = "macos")]
use tauri::menu::AboutMetadata;
use tauri::{AppHandle, Manager, Runtime};

pub fn build<R: Runtime>(app: &AppHandle<R>) -> tauri::Result<Menu<R>> {
    let open = MenuItemBuilder::new("Open Game…")
        .id("open-game")
        .accelerator("CmdOrCtrl+O")
        .build(app)?;
    let library = MenuItemBuilder::new("Back to Library")
        .id("library")
        .accelerator("CmdOrCtrl+Shift+L")
        .build(app)?;

    let save = MenuItemBuilder::new("Save…")
        .id("save")
        .accelerator("CmdOrCtrl+S")
        .build(app)?;
    let restore = MenuItemBuilder::new("Restore…")
        .id("restore")
        .accelerator("CmdOrCtrl+L")
        .build(app)?;
    let restart = MenuItemBuilder::new("Restart Game")
        .id("restart")
        .accelerator("CmdOrCtrl+Shift+R")
        .build(app)?;
    let achievements = MenuItemBuilder::new("Achievements")
        .id("achievements")
        .accelerator("CmdOrCtrl+Shift+A")
        .build(app)?;

    let sidebar = MenuItemBuilder::new("Toggle Sidebar")
        .id("toggle-sidebar")
        .accelerator("CmdOrCtrl+Backslash")
        .build(app)?;
    let panel = MenuItemBuilder::new("Toggle Achievements Panel")
        .id("toggle-panel")
        .accelerator("CmdOrCtrl+I")
        .build(app)?;
    let zoom_in = MenuItemBuilder::new("Larger Text")
        .id("zoom-in")
        .accelerator("CmdOrCtrl+Equal")
        .build(app)?;
    let zoom_out = MenuItemBuilder::new("Smaller Text")
        .id("zoom-out")
        .accelerator("CmdOrCtrl+Minus")
        .build(app)?;
    let zoom_reset = MenuItemBuilder::new("Actual Size")
        .id("zoom-reset")
        .accelerator("CmdOrCtrl+Digit0")
        .build(app)?;
    let focus = MenuItemBuilder::new("Focus Mode")
        .id("toggle-focus")
        .accelerator("CmdOrCtrl+Shift+F")
        .build(app)?;
    let settings = MenuItemBuilder::new("Settings…")
        .id("settings")
        .accelerator("CmdOrCtrl+Comma")
        .build(app)?;

    let mut items: Vec<&dyn tauri::menu::IsMenuItem<R>> = Vec::new();

    // The application menu is macOS only; on Windows and Linux these live
    // under File and Help instead, which is where those platforms look.
    #[cfg(target_os = "macos")]
    let app_menu = SubmenuBuilder::new(app, "ChoiceScript")
        .about(Some(AboutMetadata {
            name: Some("ChoiceScript".into()),
            comments: Some("Games and saves stay on this machine.".into()),
            ..Default::default()
        }))
        .separator()
        .item(&settings)
        .separator()
        .services()
        .separator()
        .hide()
        .hide_others()
        .show_all()
        .separator()
        .quit()
        .build()?;
    #[cfg(target_os = "macos")]
    items.push(&app_menu);

    let file = SubmenuBuilder::new(app, "File")
        .item(&open)
        .item(&library)
        .separator()
        .item(&PredefinedMenuItem::close_window(app, None)?)
        .build()?;
    items.push(&file);

    let game = SubmenuBuilder::new(app, "Game")
        .item(&save)
        .item(&restore)
        .separator()
        .item(&restart)
        .item(&achievements)
        .build()?;
    items.push(&game);

    let view = SubmenuBuilder::new(app, "View")
        .item(&sidebar)
        .item(&panel)
        .item(&focus)
        .separator()
        .item(&zoom_in)
        .item(&zoom_out)
        .item(&zoom_reset)
        .separator()
        .item(&PredefinedMenuItem::fullscreen(app, None)?)
        .build()?;
    items.push(&view);

    #[cfg(not(target_os = "macos"))]
    let tools = SubmenuBuilder::new(app, "Tools").item(&settings).build()?;
    #[cfg(not(target_os = "macos"))]
    items.push(&tools);

    let window = SubmenuBuilder::new(app, "Window")
        .item(&PredefinedMenuItem::minimize(app, None)?)
        .item(&PredefinedMenuItem::maximize(app, None)?)
        .build()?;
    items.push(&window);

    Menu::with_items(app, &items)
}

/// The items that only mean something while a game is running.
///
/// Disabled on the library page rather than left enabled and inert: the menu is
/// where a reader looks to find out what is possible right now, and offering
/// Restart with no game open is a lie about the state of the app.
const GAME_ITEMS: &[&str] = &[
    "save",
    "restore",
    "restart",
    "achievements",
    "toggle-panel",
    "toggle-focus",
    "library",
];

#[tauri::command]
pub fn set_game_menu_enabled(app: AppHandle, enabled: bool) -> Result<(), String> {
    let Some(menu) = app.menu() else {
        return Ok(());
    };
    for id in GAME_ITEMS {
        if let Some(item) = menu.get(*id) {
            if let Some(item) = item.as_menuitem() {
                item.set_enabled(enabled).map_err(|e| e.to_string())?;
            }
        }
    }
    Ok(())
}

/// Hide the menu bar entirely, for focus mode.
///
/// The bar belongs to the window, so the front end cannot reach it — and
/// leaving it in place made focus mode a half measure. macOS draws its menu in
/// the system bar, which hides itself in fullscreen, so there is nothing to do
/// there.
#[tauri::command]
pub fn set_menu_visible(app: AppHandle, visible: bool) -> Result<(), String> {
    #[cfg(not(target_os = "macos"))]
    {
        use tauri::Manager;
        if let Some(window) = app.get_webview_window("main") {
            if visible {
                window.show_menu().map_err(|e| e.to_string())?;
            } else {
                window.hide_menu().map_err(|e| e.to_string())?;
            }
        }
    }
    #[cfg(target_os = "macos")]
    let _ = (app, visible);
    Ok(())
}
